dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/grid/postprocessor/OrgUnitColumnProcessor.java

```diff
@@ -0,0 +1,167 @@
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.common.GridHeader;
+import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.organisationunit.OrganisationUnitStore;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Luciano Fiandesio
+ */
+@Component
+public class OrgUnitColumnProcessor implements ColumnProcessor
+{
+    private final OrganisationUnitStore organisationUnitStore;
+
+    public OrgUnitColumnProcessor( OrganisationUnitStore organisationUnitStore )
+    {
+        checkNotNull( organisationUnitStore );
+
+        this.organisationUnitStore = organisationUnitStore;
+    }
+
+    @Override
+    public List<List<Object>> process( List<GridHeader> headers, List<List<Object>> rows )
+    {
+        // based on the Grid headers, fetch the Grid's indexes for the ORGANISATION_UNIT type
+        int[] indexes = getGridIndexesByType( headers, ValueType.ORGANISATION_UNIT );
+
+        if ( indexes.length > 0 )
+        {
+            // Collect all the Org Units UIDs.
+            Map<Integer, Map<Integer, String>> orgUnitUid = collectUids( indexes, rows );
+
+            if ( !orgUnitUid.isEmpty() )
+            {
+                // Map the Org Units UID to the Org Unit names
+                Map<String, String> orgUnitUidNameMap = organisationUnitStore.getOrganisationUnitUidNameMap(
```

Comment:
This might be very heavy, as it seems we might pull in all org units into memory? There might be 100k+ org units in some instances.

Reply:
This method only fetches (`in` clause) the name of OU for the UIDs of OU which are in the resultset of the query (page size in normally 100).

Reply:
Ah right okay.